File: library/src/main/java/com/afollestad/aesthetic/views/AestheticBottomNavigationView.java
package com.afollestad.aesthetic.views;

import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.support.annotation.ColorInt;
import android.support.annotation.RestrictTo;
import android.support.design.widget.BottomNavigationView;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;

import com.afollestad.aesthetic.Aesthetic;
import com.afollestad.aesthetic.BottomNavBgMode;
import com.afollestad.aesthetic.BottomNavIconTextMode;
import com.afollestad.aesthetic.R;
import com.afollestad.aesthetic.Util;

import rx.Observable;
import rx.Subscription;
import rx.subscriptions.CompositeSubscription;

import static android.support.annotation.RestrictTo.Scope.LIBRARY_GROUP;
import static com.afollestad.aesthetic.Rx.distinctToMainThread;
import static com.afollestad.aesthetic.Rx.onErrorLogAndRethrow;

/** @author Aidan Follestad (afollestad) */
@RestrictTo(LIBRARY_GROUP)
@SuppressWarnings("RestrictedApi")
public class AestheticBottomNavigationView extends BottomNavigationView {

  private Subscription modesSubscription;
  private CompositeSubscription colorSubscriptions;
  private int lastTextIconColor;

  public AestheticBottomNavigationView(Context context) {
    super(context);
  }

  public AestheticBottomNavigationView(Context context, AttributeSet attrs) {
    super(context, attrs);
  }

  public AestheticBottomNavigationView(Context context, AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);
  }

  private void invalidateIconTextColor(int backgroundColor, int selectedColor) {
    int baseColor =
        ContextCompat.getColor(
            getContext(),
            Util.isColorLight(backgroundColor) ? R.color.ate_icon_light : R.color.ate_icon_dark);
    int unselectedIconTextColor = Util.adjustAlpha(baseColor, .87f);
    ColorStateList iconColor =
        new ColorStateList(
            new int[][] {
              new int[] {-android.R.attr.state_checked}, new int[] {android.R.attr.state_checked}
            },
            new int[] {unselectedIconTextColor, selectedColor});
    ColorStateList textColor =
        new ColorStateList(
            new int[][] {
              new int[] {-android.R.attr.state_checked}, new int[] {android.R.attr.state_checked}
            },
            new int[] {unselectedIconTextColor, selectedColor});
    setItemIconTintList(iconColor);
    setItemTextColor(textColor);
  }

  @Override
  public void setBackgroundColor(@ColorInt int color) {
    super.setBackgroundColor(color);
    if (lastTextIconColor == Color.TRANSPARENT) {
      lastTextIconColor = Util.isColorLight(color) ? Color.BLACK : Color.WHITE;
    }
    invalidateIconTextColor(color, lastTextIconColor);
  }

  private void onState(State state) {
    if (colorSubscriptions != null) {
      colorSubscriptions.unsubscribe();
    }
    colorSubscriptions = new CompositeSubscription();

    switch (state.iconTextMode) {
      case BottomNavIconTextMode.SELECTED_PRIMARY:
        colorSubscriptions.add(
            Aesthetic.get()
                .colorPrimary()
                .compose(distinctToMainThread())
                .subscribe(color -> lastTextIconColor = color, onErrorLogAndRethrow()));
        break;
      case BottomNavIconTextMode.SELECTED_ACCENT:
        colorSubscriptions.add(
            Aesthetic.get()
                .colorAccent()
                .compose(distinctToMainThread())
                .subscribe(color -> lastTextIconColor = color, onErrorLogAndRethrow()));
        break;
      case BottomNavIconTextMode.BLACK_WHITE_AUTO:
        // We will automatically set the icon/text color when the background color is set
        lastTextIconColor = Color.TRANSPARENT;
        break;
      default:
        throw new IllegalStateException("Unknown bottom nav icon/text mode: " + state.iconTextMode);
    }

    switch (state.bgMode) {
      case BottomNavBgMode.PRIMARY:
        colorSubscriptions.add(
            Aesthetic.get()
                .colorPrimary()
                .compose(distinctToMainThread())
                .subscribe(this::setBackgroundColor, onErrorLogAndRethrow()));
        break;
      case BottomNavBgMode.PRIMARY_DARK:
        colorSubscriptions.add(
            Aesthetic.get()
                .colorStatusBar()
                .compose(distinctToMainThread())
                .subscribe(this::setBackgroundColor, onErrorLogAndRethrow()));
        break;
      case BottomNavBgMode.ACCENT:
        colorSubscriptions.add(
            Aesthetic.get()
                .colorAccent()
                .compose(distinctToMainThread())
                .subscribe(this::setBackgroundColor, onErrorLogAndRethrow()));
        break;
      case BottomNavBgMode.BLACK_WHITE_AUTO:
        setBackgroundColor(
            ContextCompat.getColor(
                getContext(),
                state.isDark
                    ? R.color.ate_bottom_nav_default_dark_bg
                    : R.color.ate_bottom_nav_default_light_bg));
        break;
      default:
        throw new IllegalStateException("Unknown bottom nav bg mode: " + state.bgMode);
    }
  }

  @Override
  protected void onAttachedToWindow() {
    super.onAttachedToWindow();
    modesSubscription =
        Observable.combineLatest(
                Aesthetic.get().bottomNavigationBackgroundMode(),
                Aesthetic.get().bottomNavigationIconTextMode(),
                Aesthetic.get().isDark(),
                State::create)
            .compose(distinctToMainThread())
            .subscribe(this::onState, onErrorLogAndRethrow());
  }

  @Override
  protected void onDetachedFromWindow() {
    modesSubscription.unsubscribe();
    colorSubscriptions.unsubscribe();
    super.onDetachedFromWindow();
  }

  private static class State {

    @BottomNavBgMode private final int bgMode;
    @BottomNavIconTextMode private final int iconTextMode;
    private final boolean isDark;

    private State(int bgMode, int iconTextMode, boolean isDark) {
      this.bgMode = bgMode;
      this.iconTextMode = iconTextMode;
      this.isDark = isDark;
    }

    static State create(
        @BottomNavBgMode int bgMode, @BottomNavIconTextMode int iconTextMode, boolean isDark) {
      return new State(bgMode, iconTextMode, isDark);
    }
  }
}

File: library/src/main/java/com/afollestad/aesthetic/views/AestheticBottomNavigationView.kt
/**
 * Designed and developed by Aidan Follestad (@afollestad)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.afollestad.aesthetic.views

import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.Color.BLACK
import android.graphics.Color.WHITE
import android.util.AttributeSet
import androidx.annotation.ColorInt
import com.afollestad.aesthetic.Aesthetic.Companion.get
import com.afollestad.aesthetic.BottomNavBgMode
import com.afollestad.aesthetic.BottomNavIconTextMode
import com.afollestad.aesthetic.R
import com.afollestad.aesthetic.utils.adjustAlpha
import com.afollestad.aesthetic.utils.combine
import com.afollestad.aesthetic.utils.color
import com.afollestad.aesthetic.utils.distinctToMainThread
import com.afollestad.aesthetic.utils.isColorLight
import com.afollestad.aesthetic.utils.plusAssign
import com.afollestad.aesthetic.utils.subscribeBackgroundColor
import com.afollestad.aesthetic.utils.subscribeTo
import com.afollestad.aesthetic.utils.unsubscribeOnDetach
import com.google.android.material.bottomnavigation.BottomNavigationView
import io.reactivex.disposables.CompositeDisposable

/** @author Aidan Follestad (afollestad) */
class AestheticBottomNavigationView(
  context: Context?,
  attrs: AttributeSet? = null
) : BottomNavigationView(context, attrs) {

  private var colorSubs: CompositeDisposable? = null
  private var lastTextIconColor: Int = 0
  private var backgroundColor: Int? = null
  private var iconTextMode = BottomNavIconTextMode.NONE

  private fun invalidateIconTextColor(
    backgroundColor: Int,
    selectedColor: Int
  ) {
    if (iconTextMode == BottomNavIconTextMode.NONE) return
    val baseColor = context.color(
        if (backgroundColor.isColorLight()) R.color.ate_icon_light
        else R.color.ate_icon_dark
    )
    val unselectedIconTextColor = baseColor.adjustAlpha(.87f)
    val iconColor = ColorStateList(
        arrayOf(
            intArrayOf(-android.R.attr.state_checked),
            intArrayOf(android.R.attr.state_checked)
        ),
        intArrayOf(unselectedIconTextColor, selectedColor)
    )
    val textColor = ColorStateList(
        arrayOf(
            intArrayOf(-android.R.attr.state_checked),
            intArrayOf(android.R.attr.state_checked)
        ),
        intArrayOf(unselectedIconTextColor, selectedColor)
    )
    itemIconTintList = iconColor
    itemTextColor = textColor
  }

  override fun setBackgroundColor(@ColorInt color: Int) {
    super.setBackgroundColor(color)
    this.backgroundColor = color
    this.itemBackground = null
    if (lastTextIconColor == Color.TRANSPARENT) {
      lastTextIconColor = if (color.isColorLight()) BLACK else WHITE
    }
    invalidateIconTextColor(color, lastTextIconColor)
  }

  private fun onState(state: State) {
    colorSubs?.clear()
    colorSubs = CompositeDisposable()
    this.iconTextMode = state.iconTextMode

    when (state.iconTextMode) {
      BottomNavIconTextMode.SELECTED_PRIMARY ->
        colorSubs += get().colorPrimary()
            .distinctToMainThread()
            .subscribeTo {
              lastTextIconColor = it
              invalidateWithBackgroundColor()
            }

      BottomNavIconTextMode.SELECTED_ACCENT ->
        colorSubs += get().colorAccent()
            .distinctToMainThread()
            .subscribeTo {
              lastTextIconColor = it
              invalidateWithBackgroundColor()
            }

      BottomNavIconTextMode.BLACK_WHITE_AUTO -> {
        // We will automatically set the icon/text color when the background color is set
        lastTextIconColor = Color.TRANSPARENT
        invalidateWithBackgroundColor()
      }

      BottomNavIconTextMode.NONE -> {
        // no-op
      }
    }

    when (state.bgMode) {
      BottomNavBgMode.PRIMARY ->
        colorSubs += get().colorPrimary()
            .distinctToMainThread()
            .subscribeBackgroundColor(this)

      BottomNavBgMode.PRIMARY_DARK ->
        colorSubs += get().colorStatusBar()
            .distinctToMainThread()
            .subscribeBackgroundColor(this)

      BottomNavBgMode.ACCENT ->
        colorSubs += get().colorAccent()
            .distinctToMainThread()
            .subscribeBackgroundColor(this)

      BottomNavBgMode.BLACK_WHITE_AUTO ->
        setBackgroundColor(
            context.color(
                if (state.isDark) R.color.ate_bottom_nav_default_dark_bg
                else R.color.ate_bottom_nav_default_light_bg
            )
        )

      BottomNavBgMode.NONE -> {
        // no-op
      }
    }
  }

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()
    combine(
        get().bottomNavigationBackgroundMode(),
        get().bottomNavigationIconTextMode(),
        get().isDark
    ) { bgMode, iconTextMode, isDark -> State(bgMode, iconTextMode, isDark) }
        .distinctToMainThread()
        .subscribeTo(::onState)
        .unsubscribeOnDetach(this)
  }

  override fun onDetachedFromWindow() {
    colorSubs?.clear()
    super.onDetachedFromWindow()
  }

  private fun invalidateWithBackgroundColor() {
    if (backgroundColor != null) {
      setBackgroundColor(backgroundColor!!)
    }
  }

  private data class State(
    val bgMode: BottomNavBgMode,
    val iconTextMode: BottomNavIconTextMode,
    val isDark: Boolean
  )
}
